module/documents/actor/actor.mjs

```diff
@@ -1716,6 +1726,31 @@ export default class Actor5e extends Actor {
       if ( recoverLongRestUses && item.system.recharge?.value ) {
         updates.push({_id: item.id, "system.recharge.charged": true});
       }
+
+      // Items that roll to gain charges on a new day
+      if ( recoverDailyUses && (item.system.uses?.per === "charges")
+           && (item.system.uses.recoveryFormula !== "")
+           && (item.system.uses.value !== item.system.uses.max) ) {
+        const roll = new CONFIG.Dice.DamageRoll(item.system.uses.recoveryFormula, this.getRollData());
+
+        if ( recoverLongRestUses && (game.settings.get("dnd5e", "restVariant") === "gritty") ) {
+          roll.alter(7, 0, {multiplyNumeric: true});
+        }
+
+        await roll.evaluate({async: true});
+        if ( roll ) {
+          const count = Math.min(item.system.uses.max - item.system.uses.value, roll.total);
+          const isMax = (item.system.uses.value + count) === item.system.uses.max;
+          const locKey = `DND5E.ItemRecoveryRoll${isMax ? "Max" : ""}`;
+          updates.push({_id: item.id, "system.uses.value": item.system.uses.value + count});
+          roll.options.flavor = game.i18n.format(locKey, { name: item.name, count });
+          rolls.push(roll);
+          await roll.toMessage({
+            user: game.user.id,
+            speaker: { actor: this, alias: this.name }
+          });
```

Comment:
I guess we could support @krbz999's suggestion without too much trouble (assuming we relax the `item.system.uses.value !== item.system.uses.max` in the the `if` guard):

```js
const uses = item.system.uses;
const newValue = Math.clamped(uses.value + roll.total, 0, uses.max);
if ( newValue !== uses.value ) {
  const diff = newValue - uses.value;
  const isMax = newValue === uses.max;
  const locKey = `DND5E.Item${diff < 0 ? "Loss" : "Recovery"}Roll${isMax ? "Max": ""}`;
  updates.push({_id: item.id, "system.uses.value": newValue});
  roll.options.flavor = game.i18n.format(locKey, {name: item.name, count: diff});
  rolls.push(roll);
  await roll.toMessage({
    user: game.user.id,
    speaker: {actor: this, alias: this.name}
  });
}
```

Reply:
I would also like to suggest that the 7 (for gritty) is not hard coded and instead moved to the CONFIG and able to be overridden.